integration-tests/smoke/ccip_batching_test.go

```diff
@@ -0,0 +1,428 @@
+
+	state, err := changeset.LoadOnchainState(e.Env)
+	require.NoError(t, err)
+
+	allChainSelectors := maps.Keys(e.Env.Chains)
+	require.Len(t, allChainSelectors, 3, "this test expects 3 chains")
+	sourceChain1 := allChainSelectors[0]
+	sourceChain2 := allChainSelectors[1]
+	destChain := allChainSelectors[2]
+	t.Log("All chain selectors:", allChainSelectors,
+		", home chain selector:", e.HomeChainSel,
+		", feed chain selector:", e.FeedChainSel,
+		", source chain selector 1:", sourceChain1,
+		", source chain selector 2:", sourceChain2,
+		", dest chain selector:", destChain,
+	)
+
+	// connect sourceChain1 and sourceChain2 to destChain
+	require.NoError(t, changeset.AddLaneWithDefaultPrices(e.Env, state, sourceChain1, destChain))
+	require.NoError(t, changeset.AddLaneWithDefaultPrices(e.Env, state, sourceChain2, destChain))
+
+	const (
+		numMessages = 5
+	)
+	var (
+		startSeqNum = map[uint64]ccipocr3.SeqNum{
+			sourceChain1: 1,
+			sourceChain2: 1,
+		}
+		endSeqNum = map[uint64]ccipocr3.SeqNum{
+			sourceChain1: ccipocr3.SeqNum(numMessages),
+			sourceChain2: ccipocr3.SeqNum(numMessages),
+		}
+	)
+
+	t.Run("batch data only messages from single source", func(t *testing.T) {
+		err := sendMessages(
+			ctx,
+			t,
+			e.Env.Chains[sourceChain1],
+			e.Env.Chains[sourceChain1].DeployerKey,
+			state.Chains[sourceChain1].OnRamp,
+			state.Chains[sourceChain1].Router,
+			state.Chains[sourceChain1].Multicall3,
+			destChain,
+			numMessages,
+			common.LeftPadBytes(state.Chains[destChain].Receiver.Address().Bytes(), 32),
+		)
+		require.NoError(t, err)
+
+		_, err = changeset.ConfirmCommitWithExpectedSeqNumRange(
+			t,
+			e.Env.Chains[sourceChain1],
+			e.Env.Chains[destChain],
+			state.Chains[destChain].OffRamp,
+			nil,
+			ccipocr3.NewSeqNumRange(startSeqNum[sourceChain1], endSeqNum[sourceChain1]),
+		)
+		require.NoErrorf(t, err, "failed to confirm commit from chain %d", sourceChain1)
+
+		states, err := changeset.ConfirmExecWithSeqNrs(
+			t,
+			e.Env.Chains[sourceChain1],
+			e.Env.Chains[destChain],
+			state.Chains[destChain].OffRamp,
+			nil,
+			genSeqNrRange(startSeqNum[sourceChain1], endSeqNum[sourceChain1]),
+		)
+		require.NoError(t, err)
+		// assert that all states are successful
+		for _, state := range states {
+			require.Equal(t, changeset.EXECUTION_STATE_SUCCESS, state)
+		}
+
+		startSeqNum[sourceChain1] = endSeqNum[sourceChain1] + 1
+		endSeqNum[sourceChain1] = startSeqNum[sourceChain1] + ccipocr3.SeqNum(numMessages) - 1
+	})
+
+	t.Run("batch data only messages from multiple sources", func(t *testing.T) {
+		var (
+			wg           sync.WaitGroup
+			sourceChains = []uint64{sourceChain1, sourceChain2}
+			errs         = make(chan error, len(sourceChains))
+		)
+
+		for _, srcChain := range sourceChains {
+			wg.Add(1)
+			go sendMessagesAsync(
+				ctx,
+				t,
+				e,
+				state,
+				srcChain,
+				destChain,
+				numMessages,
+				&wg,
+				errs,
+			)
+		}
+
+		wg.Wait()
+
+		var i int
+	outer:
```

Comment:
This is duplicated 3 times, please add a function that takes the channel as an input

Reply:
I think having the function here is not as good, there's a ton of vars to take in. Will have to think about how to refactor this to be prettier later

Reply:
It's 3 variables
* testing.T
* errs channel
* sourceChains length

Much better IMO than current.

Reply:
The chan types are all different unfortunately, there is `chan error`, `chan outputErr[CommitReport]` and `chan outputErr[map[...]]`